pkg/workspace/resource/node_claim.go

```diff
@@ -48,179 +49,203 @@ func NewNodeClaimManager(c client.Client, recorder record.EventRecorder, expecta
-					"NodeClaimNotReady", fmt.Sprintf("NodeClaim %s is not ready yet", nodeClaim.Name)); updateErr != nil {
-					klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
-				}
-
-				return false, nil
-			}
+	// Sort NodeClaims for deletion: deletion timestamp set first, then not ready ones,
+	// then by creation timestamp (newest first)
+	sort.Slice(claimsWithoutPods, func(i, j int) bool {
+		nodeClaimI := claimsWithoutPods[i]
+		nodeClaimJ := claimsWithoutPods[j]
+
+		deletingI := nodeClaimI.DeletionTimestamp != nil
+		deletingJ := nodeClaimJ.DeletionTimestamp != nil
+		if deletingI != deletingJ {
+			return deletingI // being deleted comes first
 		}
-		// All NodeClaims are ready - update status condition to indicate success
-		if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionTrue,
-			"NodeClaimsReady", fmt.Sprintf("All NodeClaims are ready (NodeClaims: %d)", currentNodeClaimCount)); updateErr != nil {
-			klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
-			return false, fmt.Errorf("failed to update NodeClaim status condition: %w", updateErr)
+
+		readyI := nodeclaim.IsNodeClaimReadyNotDeleting(nodeClaimI)
+		readyJ := nodeclaim.IsNodeClaimReadyNotDeleting(nodeClaimJ)
+		if readyI != readyJ {
+			return !readyI // not ready comes first (true when i is not ready)
 		}
-		return true, nil
-	}
 
-	return false, nil
-}
+		return nodeClaimI.CreationTimestamp.After(nodeClaimJ.CreationTimestamp.Time)
+	})
+
+	for _, nodeClaim := range claimsWithoutPods[:claimsToDelete] {
+		if nodeClaim.DeletionTimestamp.IsZero() {
+			if err := c.Client.Delete(ctx, nodeClaim); err != nil {
+				c.expectations.DeletionObserved(c.logger, workspaceKey)
+				klog.ErrorS(err, "failed to delete NodeClaim",
+					"nodeClaim", nodeClaim.Name,
+					"workspace", workspaceKey)
+				c.recorder.Eventf(wObj, "Warning", "NodeClaimDeletionFailed",
+					"Failed to delete NodeClaim %s for workspace %s: %v", nodeClaim.Name, wObj.Name, err)
+				continue // should not return here or expectations will leak
+			}
+			klog.InfoS("NodeClaim deleted successfully",
+				"nodeClaim", nodeClaim.Name,
+				"creationTimestamp", nodeClaim.CreationTimestamp,
+				"workspace", workspaceKey)
 
-// isNodeClaimReady checks if a NodeClaim is in ready state
-func (c *NodeClaimManager) isNodeClaimReady(nodeClaim *karpenterv1.NodeClaim) bool {
-	for _, condition := range nodeClaim.Status.Conditions {
-		if condition.Type == "Ready" {
-			return condition.Status == "True"
+			c.recorder.Eventf(wObj, "Normal", "NodeClaimDeleted",
+				"Successfully deleted NodeClaim %s for workspace %s", nodeClaim.Name, wObj.Name)
+		} else {
+			c.expectations.DeletionObserved(c.logger, workspaceKey)
 		}
 	}
 
-	return nodeClaim.Status.NodeName != ""
+	return true, nil
```

Comment:
is the bool variable here really necessary? 
what about simply return error

Reply:
fixed